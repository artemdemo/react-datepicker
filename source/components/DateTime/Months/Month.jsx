import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { propIsMoment } from '../propTypes';

import './Month.less';

class Month extends React.Component {
    clickHandler(validDate) {
        const { date, onClick } = this.props;
        if (validDate) {
            onClick && onClick(date);
        }
    }

    render() {
        const { date, current, selected, faded, isValidDate } = this.props;
        const monthClass = classnames({
            'datetime-month': true,
        });
        const validDate = isValidDate ? isValidDate(date) : true;
        const contentClass = classnames({
            'datetime-month__content': true,
            'datetime-month__content_current': current,
            'datetime-month__content_selected': selected,
            'datetime-month__content_faded': faded,
            'datetime-month__content_disabled': !validDate,
        });
        return (
            <div
                className={monthClass}
                onClick={this.clickHandler.bind(this, validDate)}
            >
                <div className={contentClass}>
                    {date.format('MMM')}
                </div>
            </div>
        );
    }
}

Month.displayName = 'Month';

Month.propTypes = {
    date: propIsMoment.isRequired,
    onClick: PropTypes.func,
    isValidDate: PropTypes.func,
    current: PropTypes.bool,
    selected: PropTypes.bool,
    faded: PropTypes.bool,
};

Month.defaultProps = {
    onClick: null,
    isValidDate: null,
    current: false,
    selected: false,
    faded: false,
};

export default Month;
