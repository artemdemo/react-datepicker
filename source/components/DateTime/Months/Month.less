@import '../styles/general';

.datetime-month {
    float: left;
    width: 25%;
    height: 67px;
    text-align: center;
    display: table;
}

    .datetime-month__content {
        .clickable();
        display: table-cell;
        vertical-align: middle;

        border: 2px solid transparent;
        background-color: transparent;
        height: 100%;
    }
        .datetime-month__content_faded {
            color: fade(@text-color, 35%);
        }
        .datetime-month__content_disabled {
            color: fade(@text-color, 15%);
            cursor: not-allowed;

            &:hover {
                background: transparent;
            }
        }
