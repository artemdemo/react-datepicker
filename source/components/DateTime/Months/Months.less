@import '../styles/general';

.datetime-months {
    .clearfix();
}
