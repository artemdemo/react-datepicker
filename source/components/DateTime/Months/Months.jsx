import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import Month from './Month';
import { propIsMoment } from '../propTypes';

import './Months.less';

class Months extends React.Component {
    monthClickHandler(date) {
        const { onDateChange } = this.props;
        onDateChange(
            date
                .clone()
                .month(date.month())
                .year(date.year()),
        );
    }

    renderMonths() {
        const { date, isValidDate } = this.props;
        const syntheticMoment = moment(date).month(0).date(1);
        const monthsList = [];

        for (let i = 0; i < 12; i++) {
            monthsList.push(
                <Month
                    date={syntheticMoment.clone()}
                    onClick={this.monthClickHandler.bind(this)}
                    isValidDate={isValidDate}
                    key={`datetime-months__month-${i}`}
                />,
            );
            syntheticMoment.add(1, 'month');
        }

        return monthsList;
    }

    render() {
        return (
            <div className='datetime-months'>
                {this.renderMonths()}
            </div>
        );
    }
}

Months.displayName = 'Months';

Months.propTypes = {
    date: propIsMoment.isRequired,
    selectedDate: propIsMoment.isRequired,
    onDateChange: PropTypes.func.isRequired,
    isValidDate: PropTypes.func,
};

Months.defaultProps = {
    isValidDate: null,
};

export default Months;
