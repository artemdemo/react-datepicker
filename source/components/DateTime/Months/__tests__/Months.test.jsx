import React from 'react';
import moment from 'moment';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Months from '../Months';

describe('<Months />', () => {
    const date = moment(1502459094000);

    it('simple render', () => {
        const tree = renderer.create(
            <Months
                date={date}
                onDateChange={() => {}}
                test
            />,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('handle click on date', () => {
        let selectedTimestamp = 0;
        const wrapper = mount(
            <Months
                date={date}
                selectedDate={date}
                onDateChange={(date) => {
                    selectedTimestamp = date.valueOf();
                }}
                test
            />,
        );
        wrapper.find('.datetime-month').first().simulate('click');
        expect(selectedTimestamp).toEqual(1483281894000);
    });

    it('all months after august (not incl) should be disabled', () => {
        const tree = renderer.create(
            <Months
                date={date}
                selectedDate={date}
                isValidDate={_date => _date.valueOf() < 1501681494000}
                onDateChange={() => {}}
                test
            />,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
