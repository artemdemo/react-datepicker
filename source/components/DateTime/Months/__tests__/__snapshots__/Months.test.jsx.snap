// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Months /> all months after august (not incl) should be disabled 1`] = `
<div
  className="datetime-months"
  style={
    Object {
      "width": 200,
    }
  }
>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Jan
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Feb
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Mar
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Apr
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      May
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Jun
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Jul
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Aug
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content datetime-month__content_disabled"
    >
      Sep
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content datetime-month__content_disabled"
    >
      Oct
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content datetime-month__content_disabled"
    >
      Nov
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content datetime-month__content_disabled"
    >
      Dec
    </div>
  </div>
</div>
`;

exports[`<Months /> simple render 1`] = `
<div
  className="datetime-months"
  style={
    Object {
      "width": 200,
    }
  }
>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Jan
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Feb
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Mar
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Apr
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      May
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Jun
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Jul
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Aug
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Sep
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Oct
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Nov
    </div>
  </div>
  <div
    className="datetime-month"
    onClick={[Function]}
  >
    <div
      className="datetime-month__content"
    >
      Dec
    </div>
  </div>
</div>
`;
